### ISSUE TITLE
Add Settings.transformUnneededArithmeticExpressions to optimise arithmetic prior to SQL generation

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#10644]
### Problem:

When emulating things like this in e.g. PostgreSQL:

```sql
SELECT a, b
FROM t
ORDER BY x
FETCH FIRST 10 ROWS WITH TIES
```

We'll get some unnecessary clauses and arithmetic, such as:

```sql
SELECT
  v0 AS a,
  v1 AS b
FROM (
  SELECT
    a AS v0,
    b AS v1,
    rank() OVER (ORDER BY x) AS rn
  FROM t
) x
WHERE rn > 0
AND rn <= (0 + 10)
ORDER BY rn
```

It's not necessary to filter for `rn > 0`, as it can never be `0`. Also, the arithmetic `0 + 10` is an unnecessary artifact, which should be avoided.

A similar thing can be said of `GENERATE_SERIES` emulations. An example from https://github.com/jOOQ/jOOQ/issues/10663:

```sql
select count(*) "c"
from (
  select 1 "v0"
  from (
    select ((0 + (level * 2)) - 2) generate_series
    from (
      select 1 as dual
      from systables
      where (tabid = 1)
    ) as dual
    connect by (level * 2) <= ((6 + 2) - 0)
  ) "t" ("a")
    join "informix"."t_book"
      on "informix"."t_book"."id" = "t"."a"
) "t"
```

There's a lot of unnecessary arithmetic which was implemented because the `GENERATE_SERIES(from, to)` arguments could have been generic `Field` expressions.

### Caveats

Optimising this (even if the arithmetic is produced internally) is risky for these reasons:

- Users might use named parameters, and they might expect the named parameter to be present in the generated SQL. Optimising this would remove parameter names, so if users produce named parameters, we must certainly not optimise them
- Existing user code might call `Query.getBindValues()` and hard code against current behaviour (e.g. make assumptions about there bing N bind values, when after this change, there might be N - 3 or whatever number of binds)

These caveats apply in general, should we optimise such expressions, not just for these *particular* emulations. So, there needs to be a setting to disable the optimisation, should we get it wrong.

### Solution

It's worth improving this more generically, but protected by a new `Settings.transformUnneededArithmeticExpressions`:

- [x] `NEVER`: status quo and default. Produce exactly the SQL that has been written by the user
- [x] `INTERNAL`: If expressions are produced from jOOQ's internals, not the user
- [x] `ALWAYS`: Always optimise expressions on bind values

Irrespective of the chosen setting, expressions that mix bind value types, e.g. `val(1).add(val(1))` and `inline(1).add(inline(1))` will be optimised to `val(2)` and `inline(2)` respectively, but `val(1).add(inline(1))` will not be optimised. Also, named parameters will never be optimised.

### Affected internal arithmetic:

- [x] Emulation of `acos()`, `asin()` for Access
- [x] Emulation of `bitOr()` for Oracle
- [x] Emulation of `cosh()`, `cot()`, `coth()`, `sinh()`, `tanh()` for most dialects
- [x] Emulation of `extract()` for a few dialects
- [x] Emulation of `for_()` loop
- [x] Emulation of `generateSeries()` for Informix, Oracle
- [x] Emulation of `lpad()` and `rpad()` for a few dialects
- [x] Emulation of `ln(value, base)` for a few dialects
- [x] Emulation of `overlay()` for most dialects
- [x] Emulation of `position()` with startindex for most dialects
- [x] Emulation of `power()` for Derby and SQLite
- [x] Emulation of `round()` for Derby
- [x] Emulation of `right()` for Derby
- [x] Emulation of `row2.overlaps(row2)` for most dialects
- [x] Emulation of `widthBucket()` for most dialects
- [x] Others, not on this list

### Affected other utilities:

- [x] `ParserCLI`
- [x] https://www.jooq.org/translate/
- [x] Documentation

### Known limitations

See: #10666

**Comment by lukaseder** [comment]
A driving motivation for this improvement is the translator (https://www.jooq.org/translate), where users may get annoyed by the excessive SQL expressions.

As mentioned in the caveats section, there may be too many users who rely on their parameters being named or otherwise identifiable in generated SQL output. For these users, a transformation (by default) might be too dangerous, especially because it may appear only in emulations on dialects for which their integration tests are not very thorough.

Better not change the default (stick to `NEVER`), and offer these optimisations as opt-in improvements (translation default can still be `IMPLICIT_BINDS`)

**Comment by lukaseder** [comment]
This will be part of the transformations implemented in #10054

**Comment by lukaseder** [comment]
Some known limitations are listed here: #10666
